Parametrizations depending on several inputs

lezcano opened the pull request:
Makes possible that the first register parametrization depends on a number of parameters rather than just one. Examples of these types of parametrizations are `torch.nn.utils.weight_norm` and low rank parametrizations via the multiplication of a `n x k`  tensor by a `k x m` tensor with `k <= m, n`.

Follows the plan outlined in https://github.com/pytorch/pytorch/pull/33344#issuecomment-768574924. A short summary of the idea is: we call `right_inverse` when registering a parametrization to generate the tensors that we are going to save. If `right_inverse` returns a sequence of tensors, then we save them as `original0`, `original1`...  If it returns a `Tensor` or a sequence of length 1, we save it as `original`.

We only allow to have many-to-one parametrizations in the first parametrization registered. The next parametrizations would need to be one-to-one.

There were a number of choices in the implementation:

If the `right_inverse` returns a sequence of parameters, then we unpack it in the forward. This is to allow to write code as:
```python
class Sum(nn.Module):
  def forward(self, X, Y):
    return X + Y
  def right_inverse(Z):
    return Z, torch.zeros_like(Z)
```
rather than having to unpack manually a list or a tuple within the `forward` function.

At the moment the errors are a bit all over the place. This is to avoid having to check some properties of `forward` and `right_inverse` when they are registered. I left this like this for now, but I believe it'd be better to call these functions when they are registered to make sure the invariants hold and throw errors as soon as possible.

The invariants are the following:
1. The following code should be well-formed
```python
X = module.weight
Y = param.right_inverse(X)
assert isinstance(Y, Tensor) or isinstance(Y, collections.Sequence)
Z = param(Y) if isisntance(Y, Tensor) else param(*Y)
```
in other words, if `Y` is a `Sequence` of `Tensor`s (we check also that the elements of the sequence are Tensors), then it is of the same length as the number parameters `param.forward` accepts.

2. Always: `X.dtype == Z.dtype and X.shape == Z.shape`. This is to protect the user from shooting themselves in the foot, as it's too odd for a parametrization to change the metadata of a tensor.
3. If it's one-to-one: `X.dtype == Y.dtype`. This is to be able to do `X.set_(Y)` so that if a user first instantiates the optimiser and then puts the parametrisation, then we reuse `X` and the user does not need to add a new parameter to the optimiser. Alas, this is not possible when the parametrisation is many-to-one. The current implementation of `spectral_norm` and `weight_norm` does not seem to care about this, so this would not be a regression. I left a warning in the documentation though, as this case is a bit tricky.

I'm still missing to go over the formatting of the documentation, I'll do that tomorrow.

lezcano committed:
Intial implementation.

albanD commented on test/test_nn.py:
```diff
@@ -2155,7 +2155,9 @@ def forward(self, X):
                 # If X is skew-symmetric it returns an orthogonal matrix
                 Id = torch.eye(X.size(0), device=X.device)
-                return torch.linalg.solve(Id + X, Id - X)
+                # We call contiguous because solve returns a tensor with strides that are Fortran-contiguous
```
Do you have a code sample that reproduces this outside of reparametrization?
Is it because you have `t.weight` that is contiguous and has a `t.weight.grad` that is also contiguous.
But then you do `t.weight.set_(other)` and that one is not while `t.weight.grad` is still contiguous?

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,63 +55,137 @@ def cached():
+            raise ValueError("'right_inverse' should return a Tensor or a Sequence of tensors (list, tuple...). "
+                             f"Got {value.__class__.__name__}")
+
+        # Set the number of original tensors
```
A common trick we use here is to get value to always be a sequence by wrapping single Tensors into a tuple of size 1.
- That simplifies the logic here as you don't need to special case for single Tensor
- You can always handle `original0` instead of `original` when there is a single Tensor
- Calling into the forward can be done with common code as well.
- If you really need to unpack it into a single Tensor (for the use of set_ for example), you can always save a boolean along with the sequence

Lezcano replied:
I tried doing that, but the treatment of these two cases is actually different enough to not be worth it---it clutters too much the implementation.

This comes from the fact that we reuse the tensor to avoid changing the `id` of the parameter and so on. Note, for example, that the exceptions raised in the if-else below are fundamentally different.

albanD reviewed (commented):
From a quick look, it looks quite good!

I think you can simplify a lot of code by making the params always a sequence as mentioned below.

Also I'm not sure to understand what prevents you from chaining two parametrizations that do have the following forward:
```
# param 1
def forward(self, p1, p2):
    return p1, p2
    
# param 2
def forward(self, p1, p2):
    return p1 + p2
```

Lezcano commented:
From offline: it's not possible to register parametrizations with several outputs, as each of these outputs would need to be a property on its own.

lezcano committed:
comment's from Alban

lezcano committed:
Check parametrizations at registration for earlier and better error
messages

lezcano committed:
minor

lezcano committed:
mypy

lezcano committed:
Errors are now mostly thrown at registration
Better testing
Better docs

Lezcano commented:
@albanD this is ready for another round of reviews.

Now the errors are all detected and thrown when registering a new parametrization. This is very good, as we test that the forward and the `right_inverse` can be run.

There used to be this very annoying problem before that, if you had made a mistake when writing your parametrization and you accessed `self.foo` and `foo` did not exist, `register_parametrization` would succeed. But then, when you called `module.weight`, since the parametrizations are registered as properties, the parametrization would raise an `AttributeError`. This error would be captured by `Module.__getattr__` and the final error that you would get was "module does not have an attribute called "weight"", which was very confusing.

This early testing should get rid of this problem and, in general, should error early, which is nice.

lezcano committed:
mypy

albanD commented on test/test_nn.py:
```diff
@@ -2416,45 +2424,230 @@ def right_inverse(self, X):
+        self.assertEqual(torch.linalg.matrix_rank(model.weight).item(), 1)
+
+        # The model should have 3 parameters now
+        for _ in range(2):
```
Why is this doing two iteration and some param update? Should this just check the number of parameters?

Lezcano replied:
I wanted to make sure that everything worked as expected. Better safe than sorry...

albanD replied:
Could you move the comments the same way you do it for the next case then? So that the comment on the number of parameter and the comment saying "this should not throw" are at the right place.
Also if you want to check interaction in a training loop, maybe use an actual optimizer here?

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
 
     Args:
-        modules (iterable): an iterable of modules representing the parametrizations
+        modules (sequence): an sequence of modules representing the parametrizations
```
nit: a sequence

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
+        #    # in the foot. This second requirement may be dropped in the future
+        #    if isinstance(Y, Tensor)
+        #      assert X.dtype == Y.dtype and X.shape == Y.shape
+        # Below we use original = X, new = Y
```
nit: why not name them original and new in your code sample?

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
+        #    Y = param.right_inverse(X)
+        #    assert isinstance(Y, Tensor) or isinstance(Y, collections.Sequence)
+        #    Z = param(Y) if isisntance(Y, Tensor) else param(*Y)
+        #    if isinstance(Y, collections.Sequence):
```
nit: you can move this check to the other assert 2 lines above no?

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
+        #    # We also require the shape to be the same to avoid the user shooting themselves
+        #    # in the foot. This second requirement may be dropped in the future
+        #    if isinstance(Y, Tensor)
+        #      assert X.dtype == Y.dtype and X.shape == Y.shape
```
I don't think we want to enforce the shape equality for X and Y. In particular, set_ will work just fine with another Tensor of a different size.

Lezcano replied:
It is not necessary, but I thought that, in the same way that we disallow to change the `dtype` for users not to encounter weird behaviours, we should also disallow changing the shape. That being said, I can remove this constraint.

Lezcano replied:
On a second thought, let's not be too conservative. I'm removing this constraint, as I could imagine it being useful.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
+                                     "(list, tuple...). "
+                                     f"Got a sequence with an element {originali.__class__.__name__}.")
+
+                # If they require grad we expect the user to add the new parametesr to the optimizer
```
I think that you would prefer to match exactly the two properties: requires_grad and Parameter or not based on the original.
Users can set requires_grad=False on Parameters, or create attributes that require gradients that are not Parameters.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,172 @@ def cached():
-            self.original.copy_(value)
+                # else we assume that right_inverse is the identity
+            if self.ntensors == 1:
+                # We know that the result is going to have the same dtype
```
I'm not sure we can make that assumption here.
You most likely want to double check things here, especially below for the number of Tensors it returns when a Sequence is involved.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -258,34 +378,86 @@ def right_inverse(self, X: Tensor) -> Tensor
+            )
+        if hasattr(parametrization, "right_inverse"):
+            X = parametrization.right_inverse(Z)  # type: ignore[operator]
+            # nb. The naming of X / Y / Z in the code vs the errors is different
```
Why does it have to be?

Lezcano replied:
https://github.com/pytorch/pytorch/pull/58488/files/eeb4f204e6fa2e36f3bee06962e3e71053735c4c#r647429832

albanD replied:
for the one above, I understand that you don't want to use original/new in the sample to keep names short.
But I'm not sure why X and Z are switched here. It would be better to have consistent naming throughout here.

albanD reviewed (commented):
You commited some submodule updates by mistake I think.

Looks good to me. Mainly some small naming updates.

lezcano committed:
Address most of Alban's comments

lezcano committed:
Fail sooner

lezcano committed:
Use a proper optimizer

Lezcano commented:
This is ready for another review @albanD 

The invariants preserved and tested for are the following:
- `model.weight` should never change of size or shape
- `right_inverse` should return a tensor or a sequence of tensors if registered on an unparametrized tensor
  - If it returns one tensor, the returned tensor should have the same dtype of the original tensor (to be able to use `set_`)
- `right_inverse` should return a tensor of the same shape and size as `model.weight` if registered on a parametrized tensor

I believe that this is a minimal way of describe the set of invariants that we need.

I hope that I have not missed any of them in the testing nor in the exceptions in the code. It was quite tricky to convince myself that everything was correct...

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,211 @@ def cached():
+            raise ValueError("'right_inverse' must return a Tensor or a Sequence of tensors (list, tuple...). "
+                             f"Got {type(new).__name__}")
+
+        # If it is a sequence of one tensor, we unpack it
```
Do we really want to do that? Also this is not documented ?

Lezcano replied:
The idea was that in the 1 input case we can reuse the original tensor, which is good. Now, this is such a weird case... What do you think we should do?

albanD replied:
If people return (foo,) instead of foo, I'm sure they have a good reason to do it. So I would lean towards removing this.
Also the way it works right now, you will pass "foo" to their forward and not "(foo,)" which is not consistent.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,211 @@ def cached():
         """
+        # All the exceptions in this function should almost never throw.
+        # They could throw if, for example, right_inverse function does not return the same dtype
+        # for every input, which should most likely be caused by a bug in the code
```
nit: maybe this could be clearer if we say that "right_inverse function returns a different dtype when given a different input"

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -53,72 +55,211 @@ def cached():
+        else:
+            originals = (getattr(self, f"original{i}") for i in range(self.ntensors))
+            x = self[0](*originals)
+        # It's not possible to call self[1:] here, so we have to be a bit more cryptic
```
nit: if you want to avoid using the internal API, you can use iter/next to drop the first element of the iterator. But I'm not sure that would be less cryptic :p

Lezcano replied:
Good point, but as you said, I don't think that it'd be much better. I think we can leave as it is.

albanD reviewed (approved):
LGTM
only small nits

lezcano committed:
Prefer sgd steps and check tha they actually work
Use collections.abc
Change the handling of a sequence of one tensor
Reword some bits

Lezcano commented:
@albanD corrected!

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -247,7 +244,7 @@ def right_inverse(self, value: Tensor) -> None:
     def forward(self) -> Tensor:
         # Unpack the originals for the first parametrization
-        if self.ntensors == 1:
+        if self.is_tensor == 1:
```
These are bool checks no? You don't want the == 1 ?

Lezcano replied:
Oooopsss. Sorry for that and thanks for the catch!

lezcano committed:
Correct == 1

walterddr commented on test/test_nn.py:
```diff
@@ -7210,7 +7495,7 @@ def test_cudnn_weight_format(self):
             weight_data = weight.data.clone()
             with torch.no_grad():
-                weight.set_(weight_data)
+                weight.copy_(weight_data)
```
this is the only place I can think of that relates to the HUD CI failure

albanD commented:
This seems to be breaking test_cudnn_weight_format in master: https://app.circleci.com/pipelines/github/pytorch/pytorch/336608/workflows/717d1325-5ae0-4b73-abf8-bced763e9dbb/jobs/14137264
I have no idea how that could be, but reverting for now.
